processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,44 +58,107 @@
  * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP or other protocols.
  * It does periodic snap shot of the list of lookup in order to bootstrap nodes after restart.
  */
-
 @ManageLifecycle
 public class LookupReferencesManager
 {
-  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
-  private final ConcurrentMap<String, LookupExtractorFactory> lookupMap = new ConcurrentHashMap<>();
-  // This is a lock against the state of the REFERENCE MANAGER (aka start/stop state), NOT of the lookup itself.
+  private static final EmittingLogger LOG = new EmittingLogger(LookupReferencesManager.class);
+
   private final ReadWriteLock startStopLock = new ReentrantReadWriteLock(true);
-  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @VisibleForTesting
+  boolean started = false;
+
+  @GuardedBy("startStopLock")
+  private final Map<String, LookupExtractorFactoryContainer> lookupMap = new HashMap<>();
 
   private final LookupSnapshotTaker lookupSnapshotTaker;
 
+  @VisibleForTesting
+  final BlockingQueue<Notice> queue = new ArrayBlockingQueue<>(10000);
+
+  private volatile ExecutorService exec;
```

Comment:
1.
I find using Executors more confusing when it is not needed, because there is only one task ever. When reader of the code sees it, he thinks "Why Executor is used here instead of Thread? Are more tasks are submitted to this executor somewhere?". With a Thread, there are no such questions.

2.
> Executors simpler than creating/managing threads explicitly

In terms of the number of calls, it's identical:
```java
exec = Executors.factoryMethod();
exec.submit(new Runnable() {});
exec.shutdownNow();
exec.awaitTermination();
```
vs.
```
thread = new Thread(new Runnable() {})
thread.start();
thread.interrupt();
thread.join();
```

3. Thread alone is lighter-weight, because ThreadPoolExecutor includes many various locks, BlockingQueue of tasks, HashSet of workers. Also `ThreadPoolExecutor` implements `finalize()`, so less such objects in the heap, better for GC pause times.